GGRC-5369 Use HEAD request in order to get ETag from BE

YuriChupyrkin opened the pull request:
<!-- If your PR depends on other tickets or PRs, you can list them here
# Dependencies

This PR is `on hold` until the dependencies are merged:

- [ ] GGRC-1234
- [ ] #1234
-->

# Issue description

Need to add possibility to send HEAD request in cases when need to get ETag only.
In that case we will avoid loading of unneeded data and improve performance

For example: we send GET request on opening Info pane, in order to get ETag, but data in response not needed on FE side

# Steps to test the changes

**First case**:
1. Open Edit modal.
_Actual result_: GET request
_Expected result_: HEAD request 

**Second case:**
1. Create new proposal;
2. Apply or decline this proposal;
_Actual result_: The app sends GET request before save (before PUT request)
_Expected result_: The app sends HEAD request before save (before PUT request)

# Solution description

1. Added new `refreshETag` method to create HEAD request.
2. Check response of _HEAD_ request and compare `ETag`s 
3. Removed unneeded using of `refresh` method. ( @reviewer, **PLEASE** check this fix. **Possible regression**)

# Sanity checklist

- [ ] I have clicked through the app to make sure my changes work and not break the app.
- [ ] I have applied the correct milestone and labels.
- [ ] My changes fix the issue described in the description (and do nothing else). 🤞
- [ ] My changes are covered by tests.
- [ ] My changes follow our [performance guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/performance.rst).
- [ ] My changes follow our [js](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/javascript.rst) and/or [python](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/python.rst) guidelines.
- [ ] My commits follow our [commit guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/how_to_write_a_commit_message.rst).

<!-- If your PR includes a migration include the additional checklist items
# Migration checklist
- [ ] Migration passes all checks from our [PR review guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/reviewing_pull_requests.rst#reviewing-a-pr-containing-database-migration-scripts)
-->

# PR Review checklist

- [ ] The changes fix the issue and don't cause any apparent regressions.
- [ ] Labels and milestone are correctly set.
- [ ] The solution description matches the changes in the code.
- [ ] There is no apparent way to improve the performance & design of the new code.
- [ ] The pull request is opened against the correct base branch.
- [ ] Upon merging, the Jira ticket's fixversion is correctly set and the ticket is moved to "QA - In Progress".

<!-- If your code is not finished yet can include a TODO check list
# TODO

- [ ] First item on the TODO
-->

iSokrat commented on src/ggrc-client/js/plugins/utils/etag-utils.js:
```diff
@@ -0,0 +1,47 @@
+  /**
+   * Gets etagData
+   * @param {string} url - resource url.
+   * @return {object} etagData - etag data.
```
Multiple `@return` definitions are a little bit strange. I suggest you to use ways described here: https://stackoverflow.com/questions/6460604/how-to-describe-object-arguments-in-jsdoc
For `@return` there is https://github.com/jsdoc3/jsdoc/issues/1045 as well.

iSokrat commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -936,6 +937,61 @@ import Mappings from './mappers/mappings';
+
+      return dfd;
+    },
+    refreshETag() {
```
This method doesn't contain any logic connected with etag. It refreshes headers via `HEAD`. Would be be better to change method's name to something else like `refreshHeaders()` etc.

iSokrat commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -936,6 +937,61 @@ import Mappings from './mappers/mappings';
+      return dfd;
+    },
+    refreshETag() {
+      let href = this.selfLink || this.href;
```
I suggest you to create separate method, that will return `href`, for example `getHref()`.

iSokrat commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -936,6 +937,61 @@ import Mappings from './mappers/mappings';
+            let dfd = this._pending_head.dfd;
+            can.ajax(ajaxOptions)
+              .done(() => {
+                dfd.resolve(href);
```
We return this `href` only for specific case - it's needed for getting `etag` from Etags store. Based on comment about method's naming, this method should be common for headers refresh. We can use proposed `getHref()` within `tryToRefreshEtag()` method for getting needed `etag`. Thus, let's remove it here.

iSokrat commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -936,6 +937,61 @@ import Mappings from './mappers/mappings';
+    tryToRefreshEtag() {
+      let dfd = can.Deferred();
+
+      this.refreshETag().then((href) => {
```
Would be better to use `getHref()` method for getting `href`.

iSokrat commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -936,6 +937,61 @@ import Mappings from './mappers/mappings';
       this._pending_refresh.fn();
       return dfd;
     },
+    tryToRefreshEtag() {
```
Could you please add a comment, that this method returns `can.Deferred(<can.Model.Cacheable>)` and how it works?

valchanin replied:
I would like to suggest name the method 'actualize', because use the method to confirm that instance version is actual.

iSokrat commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -936,6 +937,61 @@ import Mappings from './mappers/mappings';
+        if (isSuccessful) {
+          dfd.resolve(this);
+        } else {
+          this.refresh().then((data) => {
```
Your `tryToRefreshEtag()` method should return refreshed instance as I see. Could you please explicitly use `instance` or `refreshedInstance` instead of `data`.

iSokrat commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -936,6 +937,61 @@ import Mappings from './mappers/mappings';
+
+      this.refreshETag().then((href) => {
+        let etagData = EtagUtils.getEtagData(href);
+        let isSuccessful = !etagData.previousEtag ||
```
Maybe, `isUnchangedResource`?:)

iSokrat commented:
Please, resolve conflict.

valchanin commented on src/ggrc-client/js/plugins/utils/etag-utils.js:
```diff
@@ -0,0 +1,47 @@
+ Licensed under http://www.apache.org/licenses/LICENSE-2.0 <see LICENSE file>
+ */
+
+class EtagUtils {
```
I believe it is not ustils.
IMHO it is "EtagStorage"

valchanin commented on src/ggrc-client/js/plugins/utils/etag-utils.js:
```diff
@@ -0,0 +1,47 @@
+   * @return {object} etagData - etag data.
+   * @return {string} etagData.etag - etag header content.
+   * @return {string} etagData.lastModified - last modified data.
+   * @return {string} etagData.previousEtag - previous etag header content.
```
Since this code does not control data, IMHO we should not write jsdoc

valchanin commented on src/ggrc-client/js/plugins/utils/etag-utils.js:
```diff
@@ -0,0 +1,47 @@
+   * @param {string} etagData.lastModified - last modified data.
+   */
+  setEtagData(url, etagData = {}) {
+    let previousEtag = this.getEtagData(url).etag;
```
It would be better to store all previous data

valchanin commented on src/ggrc-client/js/plugins/utils/etag-utils.js:
```diff
@@ -0,0 +1,47 @@
+    this.etags[url] = etagData;
+
+    // set prev etag
+    this.etags[url].previousEtag = previousEtag;
```
It is not good to modify data that should be stored.

YuriChupyrkin commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -937,6 +941,58 @@ import {delayLeavingPageUntil} from '../plugins/utils/current-page-utils';
       return dfd;
     },
+    actualize() {
+      let dfd = can.Deferred();
```
Some pieces of source code use `actualize` instead of `refresh` and expect old API (deferred, but **not** promises).

iSokrat replied:
`Can you show me which peaces of source code expect a deferred and aren't promise compatible?`
@Smotko https://github.com/YuriChupyrkin/ggrc-core/blob/b1dbce6544aff5aa02c5f89903cc43c0083e941d/src/ggrc-client/js/models/service-models/comment.js#L53 Promises don't have `fail` method. But, as far as I see, your code will also work fine (`Also, if you take a look at my code snippet you'll see that the function actually returns a deferred.`).

YuriChupyrkin commented on src/ggrc-client/js/models/cacheable.js:
```diff
@@ -937,6 +941,58 @@ import {delayLeavingPageUntil} from '../plugins/utils/current-page-utils';
+        let href = this.getHref();
+        let etag = EtagStorage.get(href).etag;
+        let previousEtag = EtagStorage.getPrevious(href).etag;
+        let isUnchangedResource = !previousEtag || previousEtag === etag;
```
Agree! I lost this case. 

I use `!previousEtag` for following cases:
 1. data loaded by queryAPI (all properties are exist, but ETAG is absent).
 2. data is not needed (we have stub without ETAG)


Fixed

YuriChupyrkin commented:
@Smotko , generation of ETag on FE side - isn't good idea, because instance has to have `updated_at` property.

But `stubs` doesn't have that property and we will need to do `get` request to get full object with `updated_at` property.

What do you think?

YuriChupyrkin commented:
@Smotko , I have updated PR. Take a look please.

Also, I propose to review new approach to get **ETags**.
We can avoid using of **HEAD** requests in case when data loaded by queryAPI has **ETags**.
This approach will reduce number of **GET** requests to server.

But it will require a lot of changes on **Server** and **Client** sides.

What do you think?

YuriChupyrkin commented:
@Smotko 
1. Fixed. I returned back `instance.refresh()` in `tree-widget-container.js`. BUT, the app sends 2 get request when user opens info-pin first time (`tree-widget-container.js` and `ca-update.js` send **GET** request). I updated `ca-update.js` and the app sends **GET** and then **HEAD** request. I will create separate Jira ticket to fix this issue (isn't connected to current PR)

YuriChupyrkin commented:
@Smotko

1. Fixed. I returned back `instance.refresh()` in **tree-widget-container.js**. BUT, the app sends 2 **GET** requests when user opens **info-pin** first time (**tree-widget-container.js** and **ca-update.js** send **GET** requests). I updated **ca-update.js** and the app sends **GET** and then **HEAD** requests. I will create separate Jira ticket to fix this issue (isn't connected to current PR).

2. `End cycle` button sends **GET** request, because `cycle` loaded by queryAPI. Do not confuse with `workflow` which loaded by **GET** request and has `ETag`

3. Fixed.

4. **Unmap button** use `actualize` in cases when `mappingType !== "Relationship"` (P.S. I didn't find that cases)

iSokrat commented:
Please, resolve conflicts.

YuriChupyrkin commented:
@Smotko ,

1. **I cannot approve this PR until we either send a single HEAD request or we at least go back to the previous 1 GET request.**

Previously, the app sent 2 **GET** request,
Currently, the app sends 1 **GET request and 1 **HEAD** request (faster than 2 **GET** requests)

2. **Why did you change the code if the change doesn't have any actual effect?**
Because that places use `refresh` just to get **ETag** and then **SAVE** or **DELETE** entity.
The app sends **HEAD** request to check **ETags** if entity has already have **ETag**.
**Otherwise**
The app sends **GET** request to get **etag**

3. **Can you explain when refresh is preferred over actualize?**
Before **SAVE** or **DELETE**


4. **It looks like we should replace all calls to refresh with actualize**
It brakes the app in cases when `refresh` is used to GET data and entity has **Etag**

YuriChupyrkin commented:
@NastassiaKiseleva , why did you add **ON HOLD** label?

NastassiaKiseleva commented:
@YuriChupyrkin Added on hold on this PR - need a tech review and confirmation on tech approach from SME / Freemasons

YuriChupyrkin commented:
@Smotko ,

We can avoid using of HEAD requests in case when data loaded by queryAPI has ETags.
This approach will reduce number of GET requests to server.

But it will require a lot of changes on Server and Client sides.

What do you think?
